credentials/tls/certprovider/meshca/builder_test.go

```diff
@@ -0,0 +1,178 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package meshca
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/credentials/tls/certprovider"
+	configpb "google.golang.org/grpc/credentials/tls/certprovider/meshca/internal/meshca_experimental"
+	"google.golang.org/grpc/internal/testutils"
+)
+
+func overrideHTTPFuncs() func() {
+	// Directly override the functions which are used to read the zone and
+	// audience instead of overriding the http.Client.
+	origReadZone := readZoneFunc
+	readZoneFunc = func(_ httpDoer) string { return "test-zone" }
+	origReadAudience := readAudienceFunc
+	readAudienceFunc = func(_ httpDoer) string { return "test-audience" }
+	return func() {
+		readZoneFunc = origReadZone
+		readAudienceFunc = origReadAudience
+	}
+}
+
+func (s) TestBuildSameConfig(t *testing.T) {
+	defer overrideHTTPFuncs()()
+
+	// We will attempt to create `cnt` number of providers. So we create a
+	// channel of the same size here, even though we expect only one ClientConn
+	// to be pushed into this channel. This makes sure that even if more than
+	// one ClientConn ends up being created, the Build() call does not block.
+	cnt := 5
+	ccChan := testutils.NewChannelWithSize(cnt)
+
+	// Override the dial func to dial a dummy MeshCA endpoint, and also push the
+	// returned ClientConn on a channel to be inspected by the test.
+	origDialFunc := grpcDialFunc
+	grpcDialFunc = func(_ string, _ ...grpc.DialOption) (*grpc.ClientConn, error) {
+		cc, err := grpc.Dial("dummy-meshca-endpoint", grpc.WithInsecure())
+		ccChan.Send(cc)
+		return cc, err
+	}
+	defer func() { grpcDialFunc = origDialFunc }()
+
+	// Parse a good config to generate a stable config which will be passed to
+	// invocations of Build().
+	inputConfig := makeJSONConfig(t, goodConfigFullySpecified)
+	builder := newPluginBuilder()
+	stableConfig, err := builder.ParseConfig(inputConfig)
+	if err != nil {
+		t.Fatalf("builder.ParseConfig(%q) failed: %v", inputConfig, err)
+	}
+
+	// Create multiple providers with the same config. All these providers must
+	// end up sharing the same ClientConn.
+	providers := []certprovider.Provider{}
+	for i := 0; i < cnt; i++ {
+		p := builder.Build(stableConfig, certprovider.Options{})
+		if p == nil {
+			t.Fatalf("builder.Build(%s) failed: %v", string(stableConfig.Canonical()), err)
+		}
+		providers = append(providers, p)
+	}
+
+	// Make sure only one ClientConn is created. Attempt to read the second
+	// ClientConn should timeout.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	val, err := ccChan.Receive(ctx)
+	if err != nil {
+		t.Fatalf("Failed to create ClientConn: %v", err)
+	}
+	testCC := val.(*grpc.ClientConn)
+	if _, err := ccChan.Receive(ctx); err != context.DeadlineExceeded {
```

Comment:
All test cases that are expecting a timeout should probably use a much smaller timeout than `defaultTestTimeout` (1s).  How about 10ms?  Please check other tests as well.  1 second doesn't seem like much but these things tend to grow over time, and become a big source of wasted time running tests.

Reply:
> I found them to be quite flaky with a smaller deadline mainly because the tests involve creating real keys and CSRs and real signing.

For positive tests, we absolutely should use large timeouts.  10 seconds is fine for anything we expect to happen in a few milliseconds, in case there is some kind of weird slowness on travis.

But for negative tests, where we expect a timeout, we should use the shortest timeouts that we think will catch the issue, even if it only catches it 1% of the time.

Reply:
Makes sense to me finally :)
I've changed the timeouts to be generous for positive tests and be very small for negative ones.